lib/dockerregistry/checks.go

```diff
@@ -245,3 +253,193 @@ func (check *ImageSizeCheck) Run() error {
 
 	return nil
 }
+
+// MKImageVulnCheck returns an instance of ImageVulnCheck which
+// checks against images that have known vulnerabilities.
+// nolint[funlen]
+func MKImageVulnCheck(
+	syncContext SyncContext,
+	newPullEdges map[PromotionEdge]interface{},
+	severityThreshold int,
+	fakeVulnProducer ImageVulnProducer,
+) *ImageVulnCheck {
+	return &ImageVulnCheck{
+		syncContext,
+		newPullEdges,
+		severityThreshold,
+		fakeVulnProducer,
+	}
+}
+
+// Run is a function of ImageVulnCheck and checks that none of the
+// images to be promoted have any severe vulnerabilities.
+// nolint[errcheck]
+func (check *ImageVulnCheck) Run() error {
+	var populateRequests PopulateRequests = func(
+		sc *SyncContext,
+		reqs chan<- stream.ExternalRequest,
+		wg *sync.WaitGroup) {
+		srcImages := make(map[PromotionEdge]interface{})
+		for edge := range check.PullEdges {
+			srcImage := PromotionEdge{
+				Digest: edge.Digest,
+			}
+			// Only check the vulnerability for the source image if it
+			// hasn't been checked already since multiple promotion
+			// edges can contain the same source image
+			if _, found := srcImages[srcImage]; found {
+				continue
+			}
+			srcImages[srcImage] = nil
+			var req stream.ExternalRequest
+			req.RequestParams = edge
+			wg.Add(1)
+			reqs <- req
+		}
+	}
+
+	// If no custom ImageVulnProducer is provided, we use the default producer
+	// which is simply a call to the Container Analysis API which lists out
+	// the vulnerability occurrences for a given image
+	var vulnProducer ImageVulnProducer
+	if check.FakeVulnProducer != nil {
+		vulnProducer = check.FakeVulnProducer
+	} else {
+		ctx := context.Background()
+		client, err := containeranalysis.NewClient(ctx)
+		if err != nil {
+			return fmt.Errorf("NewClient: %v", err)
+		}
+		defer client.Close()
+		vulnProducer = mkRealVulnProducer(client)
+	}
+
+	vulnerableImages := make([]string, 0)
+	var processRequest ProcessRequest = func(
+		sc *SyncContext,
+		reqs chan stream.ExternalRequest,
+		requestResults chan<- RequestResult,
+		wg *sync.WaitGroup,
+		mutex *sync.Mutex) {
+		for req := range reqs {
+			reqRes := RequestResult{Context: req}
+			errors := make(Errors, 0)
+			edge := req.RequestParams.(PromotionEdge)
+			occurrences, err := vulnProducer(edge)
+			if err != nil {
+				errors = append(errors, Error{
+					Context: "error getting vulnerabilities",
+					Error:   err})
+			}
+			if len(occurrences) > 0 {
```

Comment:
I don't think you need this boolean check, because the `for` loop will only run if occurrences > 0. Right?

Reply:
Done